File: frontend/src/types/interview.ts
export type PracticeStep = "intro" | "playing" | "recording" | "review" | "report";

export interface InterviewQuestion {
  question_id: string;
  question: string;
  difficulty: string;
  question_number: number;
  total_questions: number;
}

export interface InterviewStartResponse {
  session_id: string;
  question: InterviewQuestion;
  message: string;
  job_description_id?: string;
}

export interface AnswerResponse {
  session_id: string;
  next_question?: InterviewQuestion;
  is_complete: boolean;
  message: string;
}

export interface InterviewReport {
  session_id: string;
  student_id?: string;
  role_id?: string;
  role_title?: string;
  overall_score: number;
  technical_score?: number | null;
  grammatical_score?: number | null;
  total_questions: number;
  strengths: string[];
  weaknesses: string[];
  detailed_scores: {
    question: string;
    answer: string;
    score: number;
    feedback: string;
  }[];
  recommendations: string[];
  difficulty_distribution?: Record<string, number> | null;
  subtopic_scores?: Record<string, number> | null;
  strongest_subtopics?: string[] | null;
  weakest_subtopics?: string[] | null;
  coverage_percentage?: number | null;
  recommended_learning_path?: string[] | null;
  performance_level?: string | null;
  hiring_recommendation?: string | null;
  interview_duration?: string | null;
  questions_attempted?: number | null;
  questions_answered?: number | null;
  topic_scores?: Record<string, number> | null;
  communication_analysis?: {
    speaking_speed_wpm: number;
    average_response_delay: number;
    filler_word_count: number;
    speech_confidence_score: number;
    recommendations: string[];
  } | null;
  learning_roadmap?: {
    subtopic: string;
    recommendations: string[];
  }[] | null;
  hiring_simulation?: {
    role: string;
    recommendation: string;
    confidence: number;
    reasoning: string;
  } | null;
  progression?: {
    score_trend: number[];
    confidence_trend: number[];
    topic_mastery_trend: number[];
    labels: string[];
    history_metrics?: {
      score_improvement: number;
      weakness_fixed: string[];
      new_weaknesses: string[];
    } | null;
  } | null;
  history?: {
    session_id: string;
    overall_score: number;
    completed_at: string;
    role_title: string;
  }[] | null;
  tab_switches?: number;
  completed_at: string;
}

export interface ReportHistoryItem {
  session_id: string;
  role_title: string;
  overall_score: number;
  completed_at: string;
  total_questions: number;
}

File: backend/schemas/interview.py
from pydantic import BaseModel
from typing import Optional, List, Dict, Any


class StartInterviewRequest(BaseModel):
    role_id: Optional[str] = None
    custom_role: Optional[str] = None
    interview_type: Optional[str] = "resume"
    topic_id: Optional[str] = None
    job_description_id: Optional[str] = None
    difficulty: Optional[str] = "medium"


class VerifyResumeJdRequest(BaseModel):
    role_id: Optional[str] = None
    custom_role: Optional[str] = None
    job_description_id: str


class SubmitAnswerRequest(BaseModel):
    session_id: str
    question_id: str
    answer: str


class QuitInterviewRequest(BaseModel):
    session_id: str


class InterviewQuestion(BaseModel):
    question_id: str
    question: str
    difficulty: str = "medium"
    question_number: int = 1
    total_questions: int = 10


class InterviewStartResponse(BaseModel):
    session_id: str
    question: InterviewQuestion
    message: str = "Interview started"


class AnswerResponse(BaseModel):
    session_id: str
    next_question: Optional[InterviewQuestion] = None
    is_complete: bool = False
    message: str = ""


class QuitInterviewResponse(BaseModel):
    session_id: str
    report_generated: bool = False
    message: str = ""


class QuestionScore(BaseModel):
    question: str
    answer: str
    score: int
    feedback: str


class InterviewReport(BaseModel):
    session_id: str
    overall_score: int
    technical_score: Optional[int] = None
    grammatical_score: Optional[int] = None
    total_questions: int
    strengths: List[str]
    weaknesses: List[str]
    detailed_scores: List[QuestionScore]
    recommendations: List[str]
    difficulty_distribution: Optional[Dict[str, int]] = None
    subtopic_scores: Optional[Dict[str, int]] = None
    strongest_subtopics: Optional[List[str]] = None
    weakest_subtopics: Optional[List[str]] = None
    coverage_percentage: Optional[int] = None
    recommended_learning_path: Optional[List[str]] = None
    completed_at: str
    performance_level: Optional[str] = None
    hiring_recommendation: Optional[str] = None
    interview_duration: Optional[str] = None
    questions_attempted: Optional[int] = None
    questions_answered: Optional[int] = None
    topic_scores: Optional[Dict[str, int]] = None
    communication_analysis: Optional[Dict[str, Any]] = None
    learning_roadmap: Optional[List[Dict[str, Any]]] = None
    hiring_simulation: Optional[Dict[str, Any]] = None
    progression: Optional[Dict[str, Any]] = None
    history: Optional[List[Dict[str, Any]]] = None
    tab_switches: Optional[int] = 0

